### PR TITLE
Replace the modTime (mtime) with ctime in snapshot hash

### PR DESCRIPTION
**Pull request by derekbit**
https://github.com/longhorn/longhorn/issues/4934

Signed-off-by: Derek Su <derek.su@suse.com>

### CODE REVIEW

File: pkg/replica/hash.go [hunk 1]
```diff
@@ -246,93 +243,141 @@ func (t *SnapshotHashJob) isSnapshotSilentlyCorrupted(checksum string) bool {
...
-	return fileInfo.ModTime().String(), nil
+	stat := fileInfo.Sys().(*syscall.Stat_t)
+	return time.Unix(int64(stat.Ctim.Sec), int64(stat.Ctim.Nsec)).String(), nil
 }
 
-func GetSnapshotHashInfoFromXattr(snapshotName string) (*SnapshotXattrHashInfo, error) {
-	xattrSnapshotHashValue := make([]byte, xattrSnapshotHashValueMax)
-	_, err := unix.Getxattr(diskutil.GenerateSnapshotDiskName(snapshotName), xattrSnapshotHashName, xattrSnapshotHashValue)
+func GetSnapshotHashInfoFromChecksumFile(snapshotName string) (*SnapshotXattrHashInfo, error) {
+	dir, err := os.Getwd()
```
Comment (innobead) [inline pkg/replica/hash.go:257]: Why `os.Getwd()` instead of `/var/lib/longhonr/..`?

File: pkg/replica/hash.go [hunk 2]
```diff
@@ -246,93 +243,141 @@ func (t *SnapshotHashJob) isSnapshotSilentlyCorrupted(checksum string) bool {
...
+	if err := f.Close(); err != nil {
+		return err
+	}
+
+	return os.Rename(tmpPath, path)
 }
 
-func DeleteSnapshotHashInfoFromXattr(snapshotName string) error {
-	return unix.Removexattr(diskutil.GenerateSnapshotDiskName(snapshotName), xattrSnapshotHashName)
+func DeleteSnapshotHashInfoFromChecksumFile(snapshotName string) error {
```
Comment (innobead) [inline pkg/replica/hash.go:328]: Rename to `DeleteSnapshotHashInfoChecksumFile`

File: pkg/replica/hash.go [hunk 3]
```diff
@@ -246,93 +243,141 @@ func (t *SnapshotHashJob) isSnapshotSilentlyCorrupted(checksum string) bool {
...
 		return err
 	}
+	defer f.Close()
+
+	if err := json.NewEncoder(f).Encode(&obj); err != nil {
+		return err
+	}
 
-	return unix.Setxattr(diskutil.GenerateSnapshotDiskName(snapshotName), xattrSnapshotHashName, xattrSnapshotHashValue, 0)
+	if err := f.Close(); err != nil {
```
Comment (innobead) [inline pkg/replica/hash.go:321]: You have `defer f.Close()` above already, why need here again?

**Review by innobead — CHANGES_REQUESTED** [review]
In general, LGTM. Just a few comments.

File: pkg/replica/hash.go [hunk 4]
```diff
@@ -255,47 +253,94 @@ func GetSnapshotChangeTime(snapshotName string) (string, error) {
 	return time.Unix(int64(stat.Ctim.Sec), int64(stat.Ctim.Nsec)).String(), nil
 }
 
-func GetSnapshotHashInfoFromXattr(snapshotName string) (*SnapshotXattrHashInfo, error) {
-	xattrSnapshotHashValue := make([]byte, xattrSnapshotHashValueMax)
-	_, err := unix.Getxattr(diskutil.GenerateSnapshotDiskName(snapshotName), xattrSnapshotHashName, xattrSnapshotHashValue)
+func GetSnapshotHashInfoFromChecksumFile(snapshotName string) (*SnapshotXattrHashInfo, error) {
+	dir, err := os.Getwd()
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot get working directory")
```
Comment (c3y1huang) [inline pkg/replica/hash.go:259]:
```suggestion
		return nil, errors.Wrap(err, "failed to get working directory when getting snapshot hash info")
```

File: pkg/replica/hash.go [hunk 5]
```diff
@@ -255,47 +253,94 @@ func GetSnapshotChangeTime(snapshotName string) (string, error) {
...
-	return info, nil
+	return &info, nil
 }
 
-func SetSnapshotHashInfoToXattr(snapshotName string, info *SnapshotXattrHashInfo) error {
-	xattrSnapshotHashValue, err := json.Marshal(&SnapshotXattrHashInfo{
+func SetSnapshotHashInfoToChecksumFile(snapshotName string, info *SnapshotXattrHashInfo) error {
+	dir, err := os.Getwd()
+	if err != nil {
+		return errors.Wrap(err, "cannot get working directory")
```
Comment (c3y1huang) [inline pkg/replica/hash.go:282]:
```suggestion
		return errors.Wrap(err, "failed to get working directory when setting snapshot hash info")
```

File: pkg/replica/hash.go (hunk 1 above)
Comment (derekbit) [reply pkg/replica/hash.go:257]: `os.Getwd()` can get the replica dir, which is the working dir of the sync-agent ([codes](https://github.com/longhorn/longhorn-engine/blob/master/app/cmd/replica.go#L156))

File: pkg/replica/hash.go (hunk 3 above)
Comment (derekbit) [reply pkg/replica/hash.go:321]: Redundant. Removed.

File: pkg/replica/hash.go (hunk 5 above)
Comment (derekbit) [reply pkg/replica/hash.go]: Updated, thank you.

File: pkg/replica/hash.go (hunk 4 above)
Comment (derekbit) [reply pkg/replica/hash.go]: Updated, thank you.

File: pkg/replica/hash.go (hunk 3 above)
Comment (derekbit) [reply pkg/replica/hash.go]: Updated, thank you.

File: pkg/replica/hash.go (hunk 2 above)
Comment (derekbit) [reply pkg/replica/hash.go]: Updated, thank you.

**Review by innobead — APPROVED** [review]

**Pull request merged**